src/Spago/Command/Fetch.purs

```diff
@@ -404,42 +404,41 @@ getTransitiveDepsFromPackageSet packageSet deps = do
     init = { packages: (Map.empty :: Map PackageName Package), errors: mempty }
 
-    mergeResults :: TransitiveDepsResult -> TransitiveDepsResult -> TransitiveDepsResult
-    mergeResults r1 r2 =
-      { packages: Map.union r1.packages r2.packages
-      , errors: r1.errors <> r2.errors
-      }
-
-    go :: Set PackageName -> PackageName -> StateT (Map PackageName (Map PackageName Package)) (Spago (FetchEnv a)) TransitiveDepsResult
-    go seen dep =
-      if (Set.member dep seen) then do
-        pure (init { errors { cycle = Set.singleton dep } })
+    -- | Cheap size check to be used in place of Set equality
+    errorSize :: TransitiveDepsResult -> Int
+    errorSize { errors: { notInPackageSet: p, notInIndex: i, cycle: c } } =
+      Set.size p + Set.size i + Set.size c
+
+    go :: Set PackageName -> PackageName -> StateT TransitiveDepsResult (Spago (FetchEnv a)) Unit
+    go seen dep = do
+      if Set.member dep seen then do
+        State.modify_ $ cycleError dep
       else do
-        cache <- State.get
-        case Map.lookup dep cache of
-          Just allDeps ->
-            pure (init { packages = allDeps })
-          Nothing ->
-            -- First look for the package in the set to get a version number out,
-            -- then use that version to look it up in the index and get the dependencies
-            case Map.lookup dep packageSet of
-              Nothing -> pure (init { errors { notInPackageSet = Set.singleton dep } })
-              Just package -> do
-                maybeDeps <- State.lift $ memoisedGetPackageDependencies dep package
-                case maybeDeps of
-                  Nothing -> pure (init { errors { notInIndex = Set.singleton dep } })
-                  Just dependenciesMap -> do
-                    -- recur here, as we need to get the transitive tree, not just the first level
-                    { packages: childDeps, errors } <-
-                      for (Map.toUnfoldable dependenciesMap :: Array (Tuple PackageName Range)) (\(Tuple d _) -> go (Set.insert dep seen) d)
-                        >>= (pure <<< foldl mergeResults init)
-                    let allDeps = Map.insert dep package childDeps
-                    when (Set.isEmpty (errors.cycle <> errors.notInIndex <> errors.notInPackageSet)) do
-                      State.modify_ $ Map.insert dep allDeps
-                    pure { packages: allDeps, errors }
+        alreadyRun <- Map.member dep <$> State.gets _.packages
+        when (not alreadyRun)
+          case Map.lookup dep packageSet of
+            Nothing -> State.modify_ $ notInPackageSetError dep
+            Just package -> do
+              maybeDeps <- State.lift $ memoisedGetPackageDependencies dep package
+              case maybeDeps of
+                Nothing -> State.modify_ $ notInIndexError dep
+                Just dependenciesMap -> do
+                  -- Compare errors before and after recursively running transitive deps
```

Comment:
I think this is the bit I'm struggling to get - we are comparing the amount of errors because if we don't add any errors then that means that the subtree is "clean" (i.e. error-free) and we can move forward?

Reply:
Yeah exactly that. I felt kinda gross writing it because it's so much like classic mutation.

Reply:
> when (Set.isEmpty (errors.cycle <> errors.notInIndex <> errors.notInPackageSet)) do

That was the equivalent before ☝️